src/developer-documentation/ibc/ibc.md

```diff
@@ -0,0 +1,237 @@
+                amount: amount,
+            })
+        }
+
+        pub fn native(amount: u128, denom: &str) -> Self {
+            Amount::Native(Coin {
+                denom: denom.to_string(),
+                amount: amount,
+            })
+        }
+    }
+
+    impl Amount {
+        pub fn denom(&self) -> String {
+            match self {
+                Amount::Native(c) => c.denom.clone(),
+                Amount::Cw20(c) => "cw20:".to_owned() + c.address.as_str(),
+            }
+        }
+
+        pub fn amount(&self) -> u128 {
+            match self {
+                Amount::Native(c) => c.amount,
+                Amount::Cw20(c) => c.amount,
+            }
+        }
+
+        /// convert the amount into u64
+        pub fn u64_amount(&self) -> Result<u64, Error> {
+            Ok(self.amount().try_into().unwrap())
+        }
+
+        pub fn is_empty(&self) -> bool {
+            match self {
+                Amount::Native(c) => c.amount == 0,
+                Amount::Cw20(c) => c.amount == 0,
+            }
+        }
+    }
+```
+## 2.3 error
+```js
+pub enum Error {
+        /// StdError
+        StdError,
+        /// PaymentError
+        PaymentError,
+        /// AdminError
+        AdminError,
+        /// #[error("Channel doesn't exist: {id}")]
+        NoSuchChannel { id: String },
+        /// #[error("Didn't send any funds")]
+        NoFunds {},
+        /// #[error("Amount larger than 2**64, not supported by ics20 packets")]
+        AmountOverflow {},
+        /// #[error("Only supports channel with ibc version ics20-1, got {version}")]
+        InvalidIbcVersion { version: String },
+        /// #[error("Only supports unordered channel")]
+        OnlyOrderedChannel {},
+        /// #[error("Insufficient funds to redeem voucher on channel")]
+        InsufficientFunds {},
+        /// #[error("Only accepts tokens that originate on this chain, not native tokens of remote chain")]
+        NoForeignTokens {},
+        /// #[error("Parsed port from denom ({port}) doesn't match packet")]
+        FromOtherPort { port: String },
+        /// #[error("Parsed channel from denom ({channel}) doesn't match packet")]
+        FromOtherChannel { channel: String },
+        /// #[error("Cannot migrate from different contract type: {previous_contract}")]
+        CannotMigrate { previous_contract: String },
+        /// #[error("Cannot migrate from unsupported version: {previous_version}")]
+        CannotMigrateVersion { previous_version: String },
+        /// #[error("Got a submessage reply with unknown id: {id}")]
+        UnknownReplyId { id: u64 },
+        /// #[error("You cannot lower the gas limit for a contract on the allow list")]
+        CannotLowerGas,
+        /// #[error("Only the governance contract can do this")]
+        Unauthorized,
+        /// #[error("You can only send cw20 tokens that have been explicitly allowed by governance")]
+        NotOnAllowList,
+    }
+
+```
```

Comment:
```suggestion
```rust
pub enum Error {
    /// StdError
    StdError,
    /// PaymentError
    PaymentError,
    /// AdminError
    AdminError,
    /// #[error("Channel doesn't exist: {id}")]
    NoSuchChannel { id: String },
    /// #[error("Didn't send any funds")]
    NoFunds {},
    /// #[error("Amount larger than 2**64, not supported by ics20 packets")]
    AmountOverflow {},
    /// #[error("Only supports channel with ibc version ics20-1, got {version}")]
    InvalidIbcVersion { version: String },
    /// #[error("Only supports unordered channel")]
    OnlyOrderedChannel {},
    /// #[error("Insufficient funds to redeem voucher on channel")]
    InsufficientFunds {},
    /// #[error("Only accepts tokens that originate on this chain, not native tokens of remote chain")]
    NoForeignTokens {},
    /// #[error("Parsed port from denom ({port}) doesn't match packet")]
    FromOtherPort { port: String },
    /// #[error("Parsed channel from denom ({channel}) doesn't match packet")]
    FromOtherChannel { channel: String },
    /// #[error("Cannot migrate from different contract type: {previous_contract}")]
    CannotMigrate { previous_contract: String },
    /// #[error("Cannot migrate from unsupported version: {previous_version}")]
    CannotMigrateVersion { previous_version: String },
    /// #[error("Got a submessage reply with unknown id: {id}")]
    UnknownReplyId { id: u64 },
    /// #[error("You cannot lower the gas limit for a contract on the allow list")]
    CannotLowerGas,
    /// #[error("Only the governance contract can do this")]
    Unauthorized,
    /// #[error("You can only send cw20 tokens that have been explicitly allowed by governance")]
    NotOnAllowList,
}
```

```

Reply:
sure